### PR TITLE
Associations

### PR DESCRIPTION
**Pull request by AndreaM2429**
Hi there! I am sending this pull request for you to review our code. Thanks in advance for your feedback!

# Milestones 🎯:

📌 Create the classes `Classroom`, `Book`, and `Rental`.

📌 Associations between `Classroom` and `Student`:

- Has-many side (a classroom has many students).
- Belongs-to side (a student belongs to a classroom).

📌  Associations between `Person` and `Book` using the intermediate class `Rental`:

- Has-many side of Book and Rental (a book has many rentals).
- Belongs-to side of Rental and Book (a rental belongs to a book).
- Has-many side of Person and Rental (a person has many rentals).
- Belongs-to side of Rental and Person (a rental belongs to a person).

## 🎖️ General requirements:

- [ ] Gitflow
- [ ] No linter errors
- [ ] Document work

> I added a file `tests.rb' that contains some code to check the functionality of the associations and decorator.
> Feel free to add your tests.

### CODE REVIEW

File: book.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class Book
+  attr_accessor :title, :author
+  attr_reader :book_rentals
+
```
Comment (zieeco) [inline book.rb]: - For better readability, the `book_rentals` could be simply called rentals. please, fix this.

File: book.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @book_rentals = []
```
Comment (zieeco) [inline book.rb]: - For better readability, the `book_rentals` could be simply called rentals. please, fix this to be `@rentals = [ ]`

File: book.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  end
+
+  def new_rental(book)
+    @book_rentals.push(book)
+    book.book_rental = self
+  end
```
Comment (zieeco) [inline book.rb]: - Hey, let's kindly make some improvements to your implementation by adding a date and person to the parameter and the book itself should be sent as a parameter to create the new rental achieving the 'has-many' association. 


```ruby
def new_rental(person, date)
    Rental.new(date, self, person)
  end
```

File: rental.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+  def book_rental=(book)
+    @books = book
+    book.book_rentals.push(self) unless book.book_rentals.include?(self)
+  end
+
+  def person_rental=(person)
+    @person = person
+    person.person_rentals.push(self) unless person.person_rentals.include?(self)
+  end
```
Comment (zieeco) [inline rental.rb]: - Hey, while you've added some methods which is great, however, this is more not consistent as your Rental should only take responsibility for pushing into to instance variable but in your case, you've added extra methods for that. you can further simplify it to:

```ruby
class Rental
  attr_accessor :date, :person, :book

  def initialize(date, person, book)
    @date = date
    @person = person
    @book = book
    person.rentals << self
    book.rentals << self
  end
end
```

And this does all the heavy lifting for you. please, let's improve on this. whcn you make this changes, be sure to update other part of your codes.

File: person.rb [hunk 5]
```diff
@@ -2,14 +2,15 @@
 
 class Person < Nameable
   attr_accessor :name, :age
-  attr_reader :id
+  attr_reader :id, :person_rentals
```
Comment (zieeco) [inline person.rb]: - Let's make the person_rentals to `rentals` to be consistent. a rental is a rental and someone would eventually be renting the book. so making it `rentals` is more readable.

File: person.rb [hunk 6]
```diff
@@ -2,14 +2,15 @@
...
 
   def initialize(age, name = 'Unknown', parent_permission: true)
     super()
     @id = Random.rand(1...1000)
     @name = name
     @age = age
     @parent_permission = parent_permission
+    @person_rentals = []
```
Comment (zieeco) [inline person.rb]: - Here now becomes `@rentals = []`

File: person.rb [hunk 7]
```diff
@@ -20,6 +21,11 @@ def correct_name
     @name
   end
 
+  def add_rental(rental)
+    @person_rentals.push(rental)
+    rental.person_rental = self
+  end
+
```
Comment (zieeco) [inline person.rb]: - Hey, let's kindly make some improvements to your implementation by adding a date and book to the parameter and the person itself should be sent as a parameter to create the new rental achieving the 'has-many' association

```ruby
def add_rental(book, date)
    Rental.new(date, book, self)
  end
```

**Review by zieeco — CHANGES_REQUESTED** [review]
Hi @AndreaM2429,




Great job so far! 👏 
You've done well, however, there are some issues that you still need to work on to go to the next project, but you are almost there!


## To highlight 🎯

- Great effort adding the associations ✔️
- Descriptive PR. ✔️
- Good commit messages :heavy_check_mark: 


## Required Changes ♻️
_Check the comments under the review._


## Optional suggestions

- N/A

**_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._**


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: book.rb (hunk 1 above)
Comment (AndreaM2429) [reply book.rb]: Done!

File: book.rb (hunk 2 above)
Comment (AndreaM2429) [reply book.rb]: Done!

File: book.rb (hunk 3 above)
Comment (AndreaM2429) [reply book.rb]: Done!

File: rental.rb (hunk 4 above)
Comment (AndreaM2429) [reply rental.rb]: Done!

File: person.rb (hunk 5 above)
Comment (AndreaM2429) [reply person.rb]: Done!

File: person.rb (hunk 6 above)
Comment (AndreaM2429) [reply person.rb]: Done!

File: person.rb (hunk 7 above)
Comment (AndreaM2429) [reply person.rb]: Done!

**Comment by AndreaM2429** [comment]
@zieeco thanks for all the feedback and improvements.

**Review by tresorsawasawa — APPROVED** [review]
# STATUS: APPROVE :heavy_check_mark: :heavy_check_mark:  :champagne: 

Hi @AndreaM2429 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 :clap: :clap: :100: 

nodding-yes

## To highlight :champagne: 
- All linters are passing :heavy_check_mark: 
- Clean code :heavy_check_mark: 
- Good PR description :heavy_check_mark: 

## Optional suggestions
_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me [@tresorsawasawa](https://github.com/tresorsawasawa) when you comment so I can receive the notification or use slack with the same name._

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**